pkg/resource/deploy/source_eval.go

```diff
@@ -1401,9 +1401,7 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 
 		// Invoke the provider's Construct RPC method.
 		options := plugin.ConstructOptions{
-			// We don't actually need to send a list of aliases to construct anymore because the engine does
-			// all alias construction.
-			Aliases:                 []resource.Alias{},
+			Aliases:                 aliases,
```

Comment:
Does this even work E2E with non-URN-based aliases?

I'm looking at the implementation of `Construct` in `provider_plugin.go`:

https://github.com/pulumi/pulumi/blob/18f354cfd4495c08bd5438cfde0165d6c94dc026/sdk/go/common/resource/plugin/provider_plugin.go#L1319-L1323

All that's doing is passing along the URNs to the `ConstructRequest`, not any of the other spec-based alias fields, because `ConstructRequest` only accepts URN-based aliases:

https://github.com/pulumi/pulumi/blob/18f354cfd4495c08bd5438cfde0165d6c94dc026/proto/pulumi/provider.proto#L371

And the provider API used by users to implement `Construct`, it's also just looking at the URNs:

https://github.com/pulumi/pulumi/blob/3c1a6f4a1245eb7d24df51449f295f3f994003d1/sdk/go/pulumi/provider.go#L90-L94

https://github.com/pulumi/pulumi/blob/18f354cfd4495c08bd5438cfde0165d6c94dc026/sdk/nodejs/provider/server.ts#L325

https://github.com/pulumi/pulumi/blob/18f354cfd4495c08bd5438cfde0165d6c94dc026/sdk/python/lib/pulumi/provider/server.py#L170

It looks like we may have a bigger issue here, that I'd like to understand better and have a plan for before we merge just this fix.

Reply:
> Does this even work E2E with non-URN-based aliases?

Ah no it won't.
I think a proper fix here is probably an overhaul of Construct to have a nested resource monitor, we need that for a load of other fixes as well.

I'll give a think on how we do that work.